Add absolute links support

joaofreires opened the pull request:
### Prepend the `site-url` in links with absolute path.

##### Problem:
If we host in the book in a sub page we lost the references for the relative root. We already have the `site-url` in configs, but we do not use it. If we want to link a page in the root of the book, we need to prepend a lot of `../` to reach the actual root, and if we  move pages these relative refs could lead us to dead links. 


##### Proposal
Add a new flag `use-site-url-as-root` in the `HtmlConfig`.
The changes should have backward compatibility, we'll need to enable the `use-site-url-as-root` flag in `[output.html]` section inside the `book.toml` file to use the new feature.

#### TODO: 

- [x] check if other type of links were affected by the changes 


Closes: #1764

joaofreires committed:
add absolute links support

joaofreires committed:
update docs

watsom27 commented:
Hey,

Is there any update on this one? This fixes the exact issue im having. Would be awesome if we could get this merged/looked at again.

Thanks

joaofreires commented:
I can fix the merge conflict. Though, I still need someone to review/merge it.

watsom27 commented:
You'd be an actual hero if you did, thank you. 

I replied on the issue thread too so hopefully someone notices.

schungx commented:
> You'd be an actual hero if you did, thank you.

Yes, this is long overdue.  I currently have to resort to a templating engine to handle absolute paths.  I use absolute paths almost exclusively as they don't require massive editing when I move a chapter up/down folder levels.

watsom27 commented:
Bump @ehuss

ehuss reviewed (commented):
Thanks for the PR!  

I think it is a little unfortunate to need to add a new config setting for this, but I don't see much of a way around it.

This will need some tests to be added.

It looks like links on the print page aren't using the adjusted links.

It also looks like there are some merge conflicts.

joaofreires committed:
requested changes

joaofreires committed:
add new tests

joaofreires commented on src/utils/mod.rs:
```diff
@@ -126,20 +126,32 @@ fn adjust_links<'a>(event: Event<'a>, path: Option<&Path>) -> Event<'a> {
             if let Some(caps) = MD_LINK.captures(&dest) {
-                fixed_link.push_str(&caps["link"]);
+                fixed_link.push_str(&caps["link"].trim_start_matches('/'));
                 fixed_link.push_str(".html");
                 if let Some(anchor) = caps.name("anchor") {
                     fixed_link.push_str(anchor.as_str());
                 }
+            } else if !fixed_link.is_empty() {
+                // prevent links with double slashes
+                fixed_link.push_str(&dest.trim_start_matches('/'));
             } else {
                 fixed_link.push_str(&dest);
             };
-            return CowStr::from(fixed_link);
+            if dest.starts_with('/') || path.is_some() {
+                if let Some(abs_url) = abs_url {
+                    fixed_link = format!(
+                        "{}/{}",
+                        abs_url.trim_end_matches('/'),
+                        &fixed_link.trim_start_matches('/')
+                    );
+                }
+            }
+            return CowStr::from(format!("{}", fixed_link));
```
I needed to add changes to this function because of leading double slashes when the `path` and `abs_url` var are defined. I've added some tests to make sure it only affects those cases.

joaofreires commented on guide/src/format/configuration/renderers.md:
```diff
@@ -159,6 +159,7 @@ The following configuration options are available:
   navigation links and script/css imports in the 404 file work correctly, even when accessing
   urls in subdirectories. Defaults to `/`. If `site-url` is set,
   make sure to use document relative links for your assets, meaning they should not start with `/`.
+- **use-site-url-as-root:** Prepend the `site_url` in links with absolute path.
```
Yes, done,

mgeisler replied:
> links with absolute path

As a user, I would wonder "what links are those"? It could be

* links generated by templates, such as those referencing style sheets and JavaScript code
* links generated between pages of the book
* links _I insert_ using Markdown syntax

joaofreires commented on src/utils/mod.rs:
```diff
@@ -196,12 +203,17 @@ pub fn new_cmark_parser(text: &str, curly_quotes: bool) -> Parser<'_, '_> {
+    curly_quotes: bool,
+    path: Option<&Path>,
+    abs_url: Option<&String>,
+) -> String {
```
Done, I also left a comment to merge those functions in the future.

mgeisler commented on src/utils/mod.rs:
```diff
@@ -190,11 +202,22 @@ pub fn new_cmark_parser(text: &str, curly_quotes: bool) -> Parser<'_, '_> {
+    text: &str,
+    curly_quotes: bool,
+    path: Option<&Path>,
+    abs_url: Option<&String>,
```
The `&String` type should normally not be used — `&str` is more flexible for the callers since they can pass in both string literals (that would simplify the tests below) and borrowed owned strings.

mgeisler commented:
> I think it is a little unfortunate to need to add a new config setting for this, but I don't see much of a way around it.

Would it not be better to add a syntax to let links use the `path_to_root` setting instead? That would be fully backwards compatible and avoid the need for a new config flag.

I'm thinking of something like

```markdown
To back to [the root]($path_to_root/index.md)
```

The `$path_to_root` part would be a relative link, so it works even if `site-url` is not set. The syntax could also be `{{ path_to_root }}` if you let handlebars render it.

JesusPerez commented:
[**site-url**](https://rust-lang.github.io/mdBook/format/configuration/renderers.html?highlight=site-url#html-renderer-options) value  in **book.toml**  did not work for me as far as I tried with mdbook v0.4.29  

- **chapter-item**  are not changed for {{toc}}  
- **<base href ** only works for 404.page 
- **{{ path_to_root }}**  for  **index.hbs**  has relative dots in hrefs

I can not to rely on **base href**  to solve all links and references use cases, **dots** in paths does not always help.
**SUMMARY.md** has to use **relative paths**  

To solve this I made some changes in:

- [src/renderer/html_handlebars/hbs_renderer.rs](https://github.com/JesusPerez/mdBook/commit/156dd5920c50efc5b3a77d6a20d794144d0ec42d)
- [src/renderer/html_handlebars/helpers/toc.rs](https://github.com/JesusPerez/mdBook/commit/31b7b9089b4bb8bdcb88d10def83fc2e60e6d101)

Changes can be seen in this fork:

https://github.com/rust-lang/mdBook/compare/master...JesusPerez:mdBook:master

Obviously to complete the absolute paths in the pages I use a variable **{{baseurl}}** as a prefix of the **href** and **src** links. 
 How to use is described here: [how to use absolute links in content](https://github.com/JesusPerez/mdBook#how-to-use-it) 

```markdown  
[Introduction]({{urlbase}}introduction.md)
![Image]({{urlbase}}image.jpg)
```

KFearsoff reviewed (changes requested):
The code looks good to me, and the tests run, but unfortunately I wasn't able to actually confirm on the test_book the repo has that this PR fixes the issue. Is there some issue introduced in the latest commits that breaks this?

joaofreires commented:
> > The code looks good to me, and the tests run, but unfortunately I wasn't able to actually confirm on the test_book the repo has that this PR fixes the issue. Is there some issue introduced in the latest commits that breaks this?
> 
> Confirmed here as well. Not working here: https://github.com/calvinbrewer/mdBook/tree/add_absolute_links_support with the configuration added as reference

Sorry If I'm saying something wrong, but I opened this PR more than a year ago.
I did not include any examples in the `test_book`. Could you please clarify what you mean by 'Not working here'? What specific aspect are you testing? The proposed solution involves adding the `site-url` as a root in absolute paths. Therefore, in your configuration example, it would be `/docs/`. If you create a Markdown link in your `test_book`, such as `[Prefix](/prefix.md)` (note it starts with `/`), it should correctly generate an HTML `a` element with the `href` attribute set to `/docs/prefix.html`.

KFearsoff commented:
>Therefore, in your configuration example, it would be `/docs/`. If you create a Markdown link in your `test_book`, such as `[Prefix](/prefix.md)` (note it starts with `/`), it should correctly generate an HTML `a` element with the `href` attribute set to `/docs/prefix.html`.

Hmm. I've tried doing this by myself, and after you wrote a response I checked again with the exact inputs you've provided. Here's a patch that I'm got for testing:

```patch
diff --git a/test_book/book.toml b/test_book/book.toml
index a305007..91e8c9d 100644
--- a/test_book/book.toml
+++ b/test_book/book.toml
@@ -9,6 +9,8 @@ edition = "2018"
 
 [output.html]
 mathjax-support = true
+site-url = "/docs/"
+use-site-url-as-root = true
 
 [output.html.playground]
 editable = true
diff --git a/test_book/src/individual/README.md b/test_book/src/individual/README.md
index 67a1e9d..fceb065 100644
--- a/test_book/src/individual/README.md
+++ b/test_book/src/individual/README.md
@@ -2,6 +2,8 @@
 
 This contains following tags:
 
+[Prefix](/prefix.md)
+
 - Headings
 - Paragraphs
 - Line breaks
```

I then run `cargo run -- serve --hostname 127.0.0.1`. My expectation is that at `http://127.0.0.1:3000/individual/index.html`, I should see a link to `http://127.0.0.1:3000/docs/prefix.html`, but I'm seeing a link to `http://127.0.0.1:3000/prefix.html`. Am I missing something?

joaofreires commented:
@KFearsoff  to view the results, it is necessary to execute the command `cargo run -- build`. This is because the server mode overrides the `site-url` configuration to `"/"` for development purposes, I presume.

You can check it here:
https://github.com/rust-lang/mdBook/blob/c1d622e56ea02cc47d22a6e8cdc70a836d96e721/src/cmd/serve.rs#L67

KFearsoff commented:
> @KFearsoff to view the results, it is necessary to execute the command `cargo run -- build`. This is because the server mode overrides the `site-url` configuration to `"/"` for development purposes, I presume.

Oh. *Ohhh*. Yeah, I can confirm this works.

This is... pretty unfortunate, though. I think `serve` overriding the setting is extremely unintuitive and does more harm than good, especially with this PR. Do you mind seeing if it's possible to remove, and can we afford making this breaking change? Because I'd really hate it if we can't.

joaofreires commented:
@KFearsoff, I believe this constitutes a reasonable usage. The serve mode is designed exclusively for development purposes and is tasked solely with hosting the book files. I agree that enhanced documentation would be beneficial in this context though.

Another aspect to consider carefully is the content of `site-url`. It's important to note that this is not limited to pathnames; it can include absolute URLs complete with hostnames and schemas. This flexibility could potentially lead to further complications when operating on a local server if we want to include the `site-url` as is.

It would be good to have more feedbacks on it before any change.

JesusPerez commented:
Just in case if it helps I updated  [my mdBook with absolute links via site-url](https://github.com/JesusPerez/mdBook) based in current **mdBook master branch**  (v0.4.36  59d371715924752a18a31e96081bfba3c15bb0a6)

I only fixed a couple of files, maybe it is not enough but it seems to work for me.

You can clone it, build and try following [how-to-use-it](https://github.com/JesusPerez/mdBook?tab=readme-ov-file#how-to-use-it) notes with [sitefix-book](https://github.com/JesusPerez/mdBook/tree/master/sitefix-book) as example. 

I have to deal with  multi-books services with auth,  for me it is critical to support absolute links. 

Let me know if I can help on this.

RyanSquared commented:
What needs to be done for this feature to be merged?

joaofreires committed:
Updates link handling to use `&str` for consistency

Refactors multiple functions to replace `Option<&String>` with `Option<&str>`
for improved readability and adherence to Rust best practices.

Simplifies `render_markdown_with_path` to remove unused `abs_url` parameter
and introduces `render_markdown_with_abs_path` for enhanced link rendering
flexibility.

Fixes test cases to align with the updated parameter types.

joaofreires committed:
refactor: clean up with cargo fmt

joaofreires committed:
Simplifies string conversion in link adjustment

JesusPerez commented:
This problem is still live — `site-url` only affects the 404 page, so books hosted in a subdirectory still get incorrect navigation, sidebar, asset and in-content links (the original report is #1764).

Since this PR has been waiting on the author for a while and the codebase has since moved to the `crates/` workspace layout, I've opened a fresh PR that reimplements your approach on current `master` and addresses the review here:

- **PR:** https://github.com/rust-lang/mdBook/pull/3139

It covers each point from your review on this PR:

- Tests added (integration + unit).
- Print page links now use the adjusted/absolute links and still fold cross-references to in-page anchors.
- No merge conflicts (clean reimplementation).
- No public-API/semver break — `site_url` is already a field on `HtmlConfig` in the new layout, so nothing public was added (`cargo semver-checks` passes).
- `if let` / `Option<&str>` style applied.
- TOML config docs updated.

The original approach here is yours — the new PR is marked as superseding this one and credits it. Happy to fold in any feedback there.

cc @ehuss
